### PR TITLE
feat: Eip 6110

### PR DESCRIPTION
**Pull request by gakonst**
Adds [EIP-6110](https://eips.ethereum.org/EIPS/eip-6110), blocked on https://github.com/paradigmxyz/reth/pull/8053

### CODE REVIEW

File: crates/interfaces/src/executor.rs [hunk 1]
```diff
@@ -88,12 +88,21 @@ pub enum BlockValidationError {
...
+    /// [EIP-7002]: https://eips.ethereum.org/EIPS/eip-7002
     #[error("failed to apply withdrawal requests contract call: {message}")]
     WithdrawalRequestsContractCall {
         /// The error message.
         message: String,
     },
+    /// Error when decoding deposit requests from receipts [EIP-6110]
+    ///
+    /// [EIP-6110]: https://eips.ethereum.org/EIPS/eip-6110
+    // TODO(gakonst): This is an RLP decoding error but we don't import alloy_rlp here,
```
Comment (DaniPopes) [inline crates/interfaces/src/executor.rs:102]: why not? It should be in reth_primitives already

File: crates/ethereum/evm/src/execute.rs [hunk 2]
```diff
@@ -209,13 +227,17 @@ where
...
-            .into())
+            .into());
         }
 
+        // Collect all EIP-6110 deposits
+        let deposits = parse_deposits_from_receipts(&receipts)?;
+
         // Collect all EIP-7685 requests
         let withdrawal_requests =
             post_block_withdrawal_requests(&self.chain_spec, block.timestamp, &mut evm)?;
```
Comment (DaniPopes) [inline crates/ethereum/evm/src/execute.rs:238]: Should accumulate, extending a vector, rather than two vecs and concatenating after

Comment (gakonst) [reply crates/ethereum/evm/src/execute.rs:238]: The withdrawal function returns a `Vec<Request>` via `Vec<Request>::decode` and I cannot break out of the iterator without collecting so not sure how to do that here, also prob not big perf given smol list

Comment (DaniPopes) [reply crates/ethereum/evm/src/execute.rs:238]: i see, what about moving the parse_deposits to after post_block and extending that list?

File: crates/ethereum/evm/src/eip6110.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
+// EIP-6110 parsing
```
Comment (onbjerg) [inline crates/ethereum/evm/src/eip6110.rs:1]:
```suggestion
//! EIP-6110 deposit requests parsing
```

File: crates/ethereum/evm/src/eip6110.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        // No need to filter for topic because there's only one event and that's the Deposit event
+        // in the deposit contract.
+        .filter(|log| log.address == MAINNET_DEPOSIT_CONTRACT_ADDRESS)
+        .map(|log| DepositEvent::decode_log(log, false))
+        .map(|res| {
+            let deposit = parse_deposit_from_log(&res?);
+            Ok(Request::DepositRequest(deposit))
+        })
+        .collect::<Result<Vec<_>, _>>()
+        // ugly
```
Comment (onbjerg) [inline crates/ethereum/evm/src/eip6110.rs:24]:
```suggestion
        // todo: this is ugly, we should clean it up
```

File: crates/ethereum/evm/src/eip6110.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        bytes pubkey,
+        bytes withdrawal_credentials,
+        bytes amount,
+        bytes signature,
+        bytes index
+    );
+}
+
+fn parse_deposit_from_log(log: &Log<DepositEvent>) -> DepositRequest {
+    // SAFETY: These `expect` https://github.com/ethereum/consensus-specs/blob/5f48840f4d768bf0e0a8156a3ed06ec333589007/solidity_deposit_contract/deposit_contract.sol#L107-L110
```
Comment (onbjerg) [inline crates/ethereum/evm/src/eip6110.rs:43]: great

File: crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs [hunk 6]
```diff
@@ -158,11 +158,11 @@ impl BundleStateWithReceipts {
     /// Transform block number to the index of block.
     fn block_number_to_index(&self, block_number: BlockNumber) -> Option<usize> {
         if self.first_block > block_number {
-            return None
+            return None;
```
Comment (onbjerg) [inline crates/storage/provider/src/bundle_state/bundle_state_with_receipts.rs:161]: grrrr

File: crates/ethereum/evm/src/eip6110.rs [hunk 7]
```diff
@@ -0,0 +1,75 @@
+//! EIP-6110 deposit requests parsing
+use alloy_consensus::Request;
+use alloy_eips::eip6110::{DepositRequest, MAINNET_DEPOSIT_CONTRACT_ADDRESS};
+use alloy_sol_types::{sol, SolEvent};
+use reth_interfaces::executor::BlockValidationError;
+use reth_primitives::Receipt;
+use revm_primitives::Log;
+
+pub(crate) fn parse_deposits_from_receipts(
```
Comment (onbjerg) [inline crates/ethereum/evm/src/eip6110.rs:9]: make it pub

File: crates/ethereum/evm/src/eip6110.rs (hunk 5 above)
Comment (Rjected) [reply crates/ethereum/evm/src/eip6110.rs:43]: nit: should be NOTE or something else, SAFETY should be reserved for unsafe {}

File: crates/ethereum/evm/src/eip6110.rs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    event DepositEvent(
+        bytes pubkey,
+        bytes withdrawal_credentials,
+        bytes amount,
+        bytes signature,
+        bytes index
+    );
+}
+
+fn parse_deposit_from_log(log: &Log<DepositEvent>) -> DepositRequest {
```
Comment (onbjerg) [inline crates/ethereum/evm/src/eip6110.rs:42]: maybe we should put the `DepositEvent` in alloy-eips for the relevant eip, and have a conversion from `DepositEvent → DepositRequest` in alloy

File: crates/ethereum/evm/src/eip6110.rs (hunk 5 above)
Comment (gakonst) [reply crates/ethereum/evm/src/eip6110.rs:43]: # Panics in the doc of the function actually

File: crates/ethereum/evm/src/eip6110.rs (hunk 8 above)
Comment (gakonst) [reply crates/ethereum/evm/src/eip6110.rs:42]: agreed, let's do it after

File: crates/ethereum/evm/src/execute.rs [hunk 9]
```diff
@@ -209,13 +209,17 @@ where
...
 
+        // Collect all EIP-6110 deposits
+        let deposits = crate::eip6110::parse_deposits_from_receipts(&receipts)?;
+
         // Collect all EIP-7685 requests
         let withdrawal_requests =
             post_block_withdrawal_requests(&self.chain_spec, block.timestamp, &mut evm)?;
-        let requests = withdrawal_requests;
+        // TODO: Does order matter?
+        let requests = [deposits, withdrawal_requests].concat();
```
Comment (onbjerg) [inline crates/ethereum/evm/src/execute.rs:222]: yes and this is the correct order

File: crates/ethereum/evm/src/execute.rs [hunk 10]
```diff
@@ -209,13 +209,17 @@ where
                 gas: GotExpected { got: cumulative_gas_used, expected: block.gas_used },
                 gas_spent_by_tx: receipts.gas_spent_by_tx()?,
             }
-            .into())
+            .into());
         }
 
+        // Collect all EIP-6110 deposits
+        let deposits = crate::eip6110::parse_deposits_from_receipts(&receipts)?;
```
Comment (shekhirin) [inline crates/ethereum/evm/src/execute.rs:216]:
```suggestion
        let deposit_requests = crate::eip6110::parse_deposits_from_receipts(&receipts)?;
```

File: crates/ethereum/evm/src/execute.rs [hunk 11]
```diff
@@ -209,13 +209,17 @@ where
...
+
         // Collect all EIP-7685 requests
         let withdrawal_requests =
             post_block_withdrawal_requests(&self.chain_spec, block.timestamp, &mut evm)?;
-        let requests = withdrawal_requests;
+        // Requests are ordered by Request ID.
+        let requests = [deposits, withdrawal_requests].concat();
```
Comment (shekhirin) [inline crates/ethereum/evm/src/execute.rs:222]:
```suggestion
        let requests = [deposit_requests, withdrawal_requests].concat();
```

File: crates/ethereum/evm/src/execute.rs [hunk 12]
```diff
@@ -209,13 +209,17 @@ where
...
         }
 
+        // Collect all EIP-6110 deposits
+        let deposits = crate::eip6110::parse_deposits_from_receipts(&receipts)?;
+
         // Collect all EIP-7685 requests
         let withdrawal_requests =
             post_block_withdrawal_requests(&self.chain_spec, block.timestamp, &mut evm)?;
-        let requests = withdrawal_requests;
+        // Requests are ordered by Request ID.
```
Comment (shekhirin) [inline crates/ethereum/evm/src/execute.rs:221]:
```suggestion
        // Requests are ordered by Request Type ID.
```

**Review by shekhirin — APPROVED** [review]
LGTM, only variable naming nits. We can move a bunch of stuff to alloy but let's merge it as is now to iterate faster.

File: crates/ethereum/evm/src/execute.rs [hunk 13]
```diff
@@ -311,7 +315,7 @@ where
                 receipts.iter(),
             ) {
                 debug!(target: "evm", %error, ?receipts, "receipts verification failed");
-                return Err(error)
+                return Err(error);
```
Comment (mattsse) [inline crates/ethereum/evm/src/execute.rs:318]: hehe
```suggestion
                return Err(error)
```

**Review by mattsse — APPROVED** [review]

**Pull request merged**